src/assets/TokensController.ts

```diff
@@ -268,6 +295,68 @@ export class TokensController extends BaseController<
     }
   }
 
+  /**
+   * Adds isERC721 field to token object
+   * (Called when a user attempts to add tokens that were previously added which do not yet had isERC721 field)
+   *
+   * @param {string} tokenAddress - The contract address of the token requiring the isERC721 field added.
+   * @returns {Promise<object>} The new token object with the added isERC721 field.
+   *
+   */
+  async updateTokenType(tokenAddress: string) {
+    const { tokens } = this.state;
+    const tokenIndex = tokens.findIndex((token) => {
+      return token.address === tokenAddress;
+    });
+    tokens[tokenIndex].isERC721 = await this._detectIsERC721(tokenAddress);
+    this.update({ tokens });
+    return Promise.resolve(tokens[tokenIndex]);
+  }
+
+  /**
+   * Detects whether or not a token is ERC-721 compatible.
+   *
+   * @param {string} tokensAddress - the token contract address.
+   *
+   */
+  async _detectIsERC721(tokenAddress: string) {
+    const checksumAddress = toChecksumHexAddress(tokenAddress);
+    // if this token is already in our contract metadata map we don't need
+    // to check against the contract
+    if (contractsMap[checksumAddress]?.erc721 === true) {
+      return Promise.resolve(true);
+    }
+    const tokenContract = await this._createEthersContract(
+      tokenAddress,
+      abiERC721,
+      this.ethersProvider,
+    );
+
+    return await tokenContract
+      .supportsInterface(ERC721_INTERFACE_ID)
+      .catch((error: Error) => {
+        console.log('error', error);
```

Comment:
This doesn't seem great :thinking: Is this expected to throw if the interface isn't supported? What is the error in that case?

Ideally we'd be able to identify the `false` case while still throwing any unexpected errors. We don't want a network error to mistakenly identify a collectible as a token.

Reply:
Good point. This is also a case where this is how its currently done in the extension, but I actually made this mess so I should probably clean it up, plus its a more straightforward fix than the EIP-1193 compliant provider.

Reply:
So unfortunately the error we "expect" here is coded "UNPREDICTABLE_GAS_LIMIT". I suspect this is because these contracts don't have the supportsInterface method implemented?
![Screen Shot 2021-07-15 at 4 11 36 PM](https://user-images.githubusercontent.com/34557516/125858409-ccbccdce-6ec0-4b5a-92f0-c86dd4deea08.png)

Reply:
Does it feel ok to return false when error.code is "UNPREDICTABLE_GAS_LIMIT"? or should we try to find a different way to check for the interface?

Reply:
... I guess :thinking: I'm somewhat confused about how we got a gas estimation error, given that I wouldn't expect this to estimate gas.

Reply:
Yeah... it is strange. I do see it consistently when adding non ERC721 tokens